raiden/network/utils.py

```diff
@@ -10,37 +11,56 @@
 
 
 def get_free_port(
-        bind_address: str = '127.0.0.1',
         initial_port: int = 0,
         socket_kind: SocketKind = SocketKind.SOCK_STREAM,
+        reliable: bool = True,
 ):
     """
     Find an unused TCP port.
-    This should not be used in misson-critical applications - a race condition may occur if
-    someone grabs the port before caller of this function has chance to use it.
+
+    Unless the `reliable` parameter is set to `True` (the default) this is prone to race
+    conditions - some other process may grab the port before the caller of this function has
+    a chance to use it.
+    When using `reliable` the port is forced into TIME_WAIT mode, ensuring that it will not be
+    considered 'free' by the OS for the next 60 seconds. This does however require that the
+    process using the port sets SO_REUSEADDR on it's sockets. Most 'server' applications do.
 
     If `initial_port` is passed the function will try to find a port as close as possible.
     Otherwise a random port is chosen by the OS.
 
-    Returns an iterator that will return an unused port on the specified interface.
+    Returns an iterator that will return unused port numbers.
     """
 
     def _port_generator():
         if initial_port == 0:
             next_port = repeat(0)
         else:
             next_port = count(start=initial_port)
-        for i in next_port:
+
+        for port_candidate in next_port:
+            # Don't inline the variable until https://github.com/PyCQA/pylint/issues/1437 is fixed
             sock = socket.socket(socket.AF_INET, socket_kind)
-            try:
-                sock.bind((bind_address, i))
-            except OSError as ex:
-                if ex.errno == errno.EADDRINUSE:
-                    continue
-            port = sock.getsockname()[1]
-            sock.close()
-            yield port
+            with closing(sock):
+                if reliable:
+                    sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+                try:
+                    sock.bind(('127.0.0.1', port_candidate))
```

Comment:
The reason I removed it is that the bind-to-zero-TIME_WAIT trick only works for specific interfaces (i.e. not `0.0.0.0`) since there you can't guarantee which ports are available on which interface. Also so far we didn't use anything else besides `127.0.0.1` in the test suite. 

Of course we could re-add the bind address and enforce that the combination `bind_address='0.0.0.0', reliable=True` isn't used.